File: ALODWebUtility/Common/ChangeSet.vb
Imports System.Text
Imports System.Xml
Imports ALOD.Core.Domain.Workflow
Imports ALOD.Data

Namespace Common

    Public Class ChangeSet
        Implements IList(Of ChangeRow)

        Protected _details As New Dictionary(Of Integer, ChangeSetDetails)
        Protected _sets As New List(Of ChangeRow)

        Public Sub Add(ByVal section As String, ByVal field As String, ByVal oldVal As String, ByVal newVal As String)

            Dim change As New ChangeRow(section, field, oldVal, newVal)
            _sets.Add(change)

        End Sub

        Public Sub GetByLogId(ByVal logId As Integer)
            _sets.Clear()
            Dim adapter As New SqlDataStore()
            adapter.ExecuteReader(AddressOf SetReader, "core_log_sp_GetChangeSetByLogId", logId)
            GetDetails()
        End Sub

        Public Sub GetByReferenceId(ByVal refId As Integer, ByVal type As ModuleType)
            _sets.Clear()
            Dim adapter As New SqlDataStore()
            adapter.ExecuteReader(AddressOf SetReader, "core_log_sp_GetChangeSetsByRefId", refId, type)
            GetDetails()
        End Sub

        Public Sub GetByUserID(ByVal userID As Integer)
            _sets.Clear()
            Dim adapter As New SqlDataStore()
            adapter.ExecuteReader(AddressOf SetReader, "core_log_sp_GetChangeSetByUserId", userID)
            GetDetialsByUserId(userID)
        End Sub

        Public Sub GetDetails()

            If (_sets.Count = 0) Then
                Exit Sub
            End If

            'get a list of unique log ids
            Dim list As New List(Of Integer)

            For Each row As ChangeRow In _sets
                If (Not list.Contains(row.Id)) Then
                    list.Add(row.Id)
                End If
            Next

            Dim buffer As New StringBuilder

            For Each id As Integer In list
                buffer.Append(id.ToString() + ",")
            Next

            buffer = buffer.Remove(buffer.Length - 1, 1)

            'now get the details for this
            Dim adapter As New SqlDataStore()
            adapter.ExecuteReader(AddressOf DetailsReader, "core_log_sp_GetChangeSetDetails", buffer.ToString())

        End Sub

        Public Sub Save(ByVal id As Integer)

            Dim xml As New XMLString("XML_Array")

            For Each item As ChangeRow In _sets
                xml.BeginElement("XMLList")
                xml.WriteAttribute("ID", id.ToString())
                xml.WriteAttribute("section", item.Section)
                xml.WriteAttribute("field", item.Field)
                xml.WriteAttribute("oldVal", item.OldVal)
                xml.WriteAttribute("newVal", item.NewVal)
                xml.EndElement()
            Next

            Dim adapter As New SqlDataStore()
            adapter.ExecuteNonQuery("core_log_sp_InsertChangeSet", xml.Value)

        End Sub

        Public Sub ToXml(ByRef parentNode As XmlNode, Optional ByVal repeatHeading As Boolean = True)

            Dim doc As XmlDocument = parentNode.OwnerDocument

            Dim section As String = ""
            Dim currentId As Integer = 0
            Dim noRepeat As Boolean = True
            Dim groupNode As XmlNode = doc.CreateElement("ChangeSet")

            For Each row As ChangeRow In _sets

                If (currentId <> row.Id And noRepeat) Then

                    currentId = row.Id
                    groupNode = doc.CreateElement("ChangeSet")

                    If (_details.ContainsKey(row.Id)) Then

                        Dim details As ChangeSetDetails = _details(row.Id)
                        noRepeat = repeatHeading
                        groupNode.Attributes.Append(doc.CreateAttribute("logId"))
                        groupNode.Attributes("logId").Value = details.LogId.ToString()

                        groupNode.Attributes.Append(doc.CreateAttribute("ActionDate"))
                        groupNode.Attributes("ActionDate").Value = details.ActionDate.ToString()

                        groupNode.Attributes.Append(doc.CreateAttribute("LastName"))
                        groupNode.Attributes("LastName").Value = details.LastName

                        groupNode.Attributes.Append(doc.CreateAttribute("FirstName"))
                        groupNode.Attributes("FirstName").Value = details.FirstName

                        groupNode.Attributes.Append(doc.CreateAttribute("Rank"))
                        groupNode.Attributes("Rank").Value = details.Rank

                        groupNode.Attributes.Append(doc.CreateAttribute("UserId"))
                        groupNode.Attributes("UserId").Value = details.UserId.ToString()

                    End If

                    parentNode.AppendChild(groupNode)

                End If

                Dim node As XmlNode = doc.CreateElement("ChangeRow")

                node.Attributes.Append(doc.CreateAttribute("Section"))
                node.Attributes("Section").Value = row.Section

                node.Attributes.Append(doc.CreateAttribute("Field"))
                node.Attributes("Field").Value = row.Field

                node.Attributes.Append(doc.CreateAttribute("OldValue"))
                node.Attributes("OldValue").Value = row.OldVal

                node.Attributes.Append(doc.CreateAttribute("NewValue"))
                node.Attributes("NewValue").Value = row.NewVal

                groupNode.AppendChild(node)

            Next

        End Sub

        Private Sub DetailsReader(ByVal adapter As SqlDataStore, ByVal reader As IDataReader)

            '0-logId, 1-actionDate, 2-userId, 3-lastName, 4-firstName, 5-rank

            Dim details As New ChangeSetDetails

            details.LogId = adapter.GetInteger(reader, 0)
            details.ActionDate = adapter.GetDateTime(reader, 1)
            details.UserId = adapter.GetInteger(reader, 2)
            details.LastName = adapter.GetString(reader, 3)
            details.FirstName = adapter.GetString(reader, 4)
            details.Rank = adapter.GetString(reader, 5)

            _details.Add(details.LogId, details)

        End Sub

        Private Sub GetDetialsByUserId(ByVal userId As Integer)
            Dim adapter As New SqlDataStore()
            adapter.ExecuteReader(AddressOf DetailsReader, "core_log_sp_GetLastChange", userId)
        End Sub

        Private Sub SetReader(ByVal adapter As SqlDataStore, ByVal reader As IDataReader)

            '0-logId, 1-section, 2-field, 3-old, 4-new, 5-actionId, 6-actionName
            Dim change As New ChangeRow

            change.Id = adapter.GetInteger(reader, 0)
            change.Section = adapter.GetString(reader, 1)
            change.Field = adapter.GetString(reader, 2)
            change.OldVal = adapter.GetString(reader, 3)
            change.NewVal = adapter.GetString(reader, 4)
            change.ActionId = adapter.GetNumber(reader, 5)
            change.ActionName = adapter.GetString(reader, 6)
            change.UserId = adapter.GetNumber(reader, 7)
            change.UserName = adapter.GetString(reader, 8)
            change.ActionDate = adapter.GetDateTime(reader, 9)

            _sets.Add(change)

        End Sub

#Region "IList"

        Public ReadOnly Property Count() As Integer Implements System.Collections.Generic.ICollection(Of ChangeRow).Count
            Get
                Return _sets.Count
            End Get
        End Property

        Public ReadOnly Property IsReadOnly() As Boolean Implements System.Collections.Generic.ICollection(Of ChangeRow).IsReadOnly
            Get
                Return False
            End Get
        End Property

        Default Public Property Item(ByVal index As Integer) As ChangeRow Implements System.Collections.Generic.IList(Of ChangeRow).Item
            Get
                Return _sets(index)
            End Get
            Set(ByVal value As ChangeRow)
                _sets(index) = value
            End Set
        End Property

        Public Sub Add(ByVal item As ChangeRow) Implements System.Collections.Generic.ICollection(Of ChangeRow).Add
            _sets.Add(item)
        End Sub

        Public Sub Clear() Implements System.Collections.Generic.ICollection(Of ChangeRow).Clear
            _sets.Clear()
        End Sub

        Public Function Contains(ByVal item As ChangeRow) As Boolean Implements System.Collections.Generic.ICollection(Of ChangeRow).Contains
            Return _sets.Contains(item)
        End Function

        Public Sub CopyTo(ByVal array() As ChangeRow, ByVal arrayIndex As Integer) Implements System.Collections.Generic.ICollection(Of ChangeRow).CopyTo
            _sets.CopyTo(array, arrayIndex)
        End Sub

        Public Function GetEnumerator() As System.Collections.Generic.IEnumerator(Of ChangeRow) Implements System.Collections.Generic.IEnumerable(Of ChangeRow).GetEnumerator
            Return _sets.GetEnumerator()
        End Function

        Public Function GetEnumerator1() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
            Return _sets.GetEnumerator()
        End Function

        Public Function IndexOf(ByVal item As ChangeRow) As Integer Implements System.Collections.Generic.IList(Of ChangeRow).IndexOf
            Return _sets.IndexOf(item)
        End Function

        Public Sub Insert(ByVal index As Integer, ByVal item As ChangeRow) Implements System.Collections.Generic.IList(Of ChangeRow).Insert
            _sets.Insert(index, item)
        End Sub

        Public Function Remove(ByVal item As ChangeRow) As Boolean Implements System.Collections.Generic.ICollection(Of ChangeRow).Remove
            Return _sets.Remove(item)
        End Function

        Public Sub RemoveAt(ByVal index As Integer) Implements System.Collections.Generic.IList(Of ChangeRow).RemoveAt
            _sets.RemoveAt(index)
        End Sub

#End Region

    End Class

End Namespace

File: ALODWebUtility.CS/Common/ChangeSet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Text;
using System.Xml;
using ALOD.Core.Domain.Workflow;
using ALOD.Data;

namespace ALODWebUtility.Common
{
    public class ChangeSet : IList<ChangeRow>
    {
        protected Dictionary<int, ChangeSetDetails> _details = new Dictionary<int, ChangeSetDetails>();
        protected List<ChangeRow> _sets = new List<ChangeRow>();

        public void Add(string section, string field, string oldVal, string newVal)
        {
            ChangeRow change = new ChangeRow(section, field, oldVal, newVal);
            _sets.Add(change);
        }

        public void GetByLogId(int logId)
        {
            _sets.Clear();
            SqlDataStore adapter = new SqlDataStore();
            adapter.ExecuteReader(SetReader, "core_log_sp_GetChangeSetByLogId", logId);
            GetDetails();
        }

        public void GetByReferenceId(int refId, ModuleType type)
        {
            _sets.Clear();
            SqlDataStore adapter = new SqlDataStore();
            adapter.ExecuteReader(SetReader, "core_log_sp_GetChangeSetsByRefId", refId, type);
            GetDetails();
        }

        public void GetByUserID(int userID)
        {
            _sets.Clear();
            SqlDataStore adapter = new SqlDataStore();
            adapter.ExecuteReader(SetReader, "core_log_sp_GetChangeSetByUserId", userID);
            GetDetailsByUserId(userID);
        }

        public void GetDetails()
        {
            if (_sets.Count == 0)
            {
                return;
            }

            //get a list of unique log ids
            List<int> list = new List<int>();

            foreach (ChangeRow row in _sets)
            {
                if (!list.Contains(row.Id))
                {
                    list.Add(row.Id);
                }
            }

            StringBuilder buffer = new StringBuilder();

            foreach (int id in list)
            {
                buffer.Append(id.ToString() + ",");
            }

            if (buffer.Length > 0)
            {
                buffer.Remove(buffer.Length - 1, 1);
            }

            //now get the details for this
            SqlDataStore adapter = new SqlDataStore();
            adapter.ExecuteReader(DetailsReader, "core_log_sp_GetChangeSetDetails", buffer.ToString());
        }

        public void Save(int id)
        {
            XMLString xml = new XMLString("XML_Array");

            foreach (ChangeRow item in _sets)
            {
                xml.BeginElement("XMLList");
                xml.WriteAttribute("ID", id.ToString());
                xml.WriteAttribute("section", item.Section);
                xml.WriteAttribute("field", item.Field);
                xml.WriteAttribute("oldVal", item.OldVal);
                xml.WriteAttribute("newVal", item.NewVal);
                xml.EndElement();
            }

            SqlDataStore adapter = new SqlDataStore();
            adapter.ExecuteNonQuery("core_log_sp_InsertChangeSet", xml.Value);
        }

        public void ToXml(ref XmlNode parentNode, bool repeatHeading = true)
        {
            XmlDocument doc = parentNode.OwnerDocument;

            string section = "";
            int currentId = 0;
            bool noRepeat = true;
            XmlNode groupNode = doc.CreateElement("ChangeSet");

            foreach (ChangeRow row in _sets)
            {
                if (currentId != row.Id && noRepeat)
                {
                    currentId = row.Id;
                    groupNode = doc.CreateElement("ChangeSet");

                    if (_details.ContainsKey(row.Id))
                    {
                        ChangeSetDetails details = _details[row.Id];
                        noRepeat = repeatHeading;
                        
                        XmlAttribute logIdAttr = doc.CreateAttribute("logId");
                        logIdAttr.Value = details.LogId.ToString();
                        groupNode.Attributes.Append(logIdAttr);

                        XmlAttribute actionDateAttr = doc.CreateAttribute("ActionDate");
                        actionDateAttr.Value = details.ActionDate.ToString();
                        groupNode.Attributes.Append(actionDateAttr);

                        XmlAttribute lastNameAttr = doc.CreateAttribute("LastName");
                        lastNameAttr.Value = details.LastName;
                        groupNode.Attributes.Append(lastNameAttr);

                        XmlAttribute firstNameAttr = doc.CreateAttribute("FirstName");
                        firstNameAttr.Value = details.FirstName;
                        groupNode.Attributes.Append(firstNameAttr);

                        XmlAttribute rankAttr = doc.CreateAttribute("Rank");
                        rankAttr.Value = details.Rank;
                        groupNode.Attributes.Append(rankAttr);

                        XmlAttribute userIdAttr = doc.CreateAttribute("UserId");
                        userIdAttr.Value = details.UserId.ToString();
                        groupNode.Attributes.Append(userIdAttr);
                    }

                    parentNode.AppendChild(groupNode);
                }

                XmlNode node = doc.CreateElement("ChangeRow");

                XmlAttribute sectionAttr = doc.CreateAttribute("Section");
                sectionAttr.Value = row.Section;
                node.Attributes.Append(sectionAttr);

                XmlAttribute fieldAttr = doc.CreateAttribute("Field");
                fieldAttr.Value = row.Field;
                node.Attributes.Append(fieldAttr);

                XmlAttribute oldValueAttr = doc.CreateAttribute("OldValue");
                oldValueAttr.Value = row.OldVal;
                node.Attributes.Append(oldValueAttr);

                XmlAttribute newValueAttr = doc.CreateAttribute("NewValue");
                newValueAttr.Value = row.NewVal;
                node.Attributes.Append(newValueAttr);

                groupNode.AppendChild(node);
            }
        }

        private void DetailsReader(SqlDataStore adapter, IDataReader reader)
        {
            //0-logId, 1-actionDate, 2-userId, 3-lastName, 4-firstName, 5-rank

            ChangeSetDetails details = new ChangeSetDetails();

            details.LogId = adapter.GetInteger(reader, 0);
            details.ActionDate = adapter.GetDateTime(reader, 1);
            details.UserId = adapter.GetInteger(reader, 2);
            details.LastName = adapter.GetString(reader, 3);
            details.FirstName = adapter.GetString(reader, 4);
            details.Rank = adapter.GetString(reader, 5);

            _details.Add(details.LogId, details);
        }

        private void GetDetailsByUserId(int userId)
        {
            SqlDataStore adapter = new SqlDataStore();
            adapter.ExecuteReader(DetailsReader, "core_log_sp_GetLastChange", userId);
        }

        private void SetReader(SqlDataStore adapter, IDataReader reader)
        {
            //0-logId, 1-section, 2-field, 3-old, 4-new, 5-actionId, 6-actionName
            ChangeRow change = new ChangeRow();

            change.Id = adapter.GetInteger(reader, 0);
            change.Section = adapter.GetString(reader, 1);
            change.Field = adapter.GetString(reader, 2);
            change.OldVal = adapter.GetString(reader, 3);
            change.NewVal = adapter.GetString(reader, 4);
            change.ActionId = (int)adapter.GetNumber(reader, 5); // GetNumber likely returns decimal or double, casting to int
            change.ActionName = adapter.GetString(reader, 6);
            change.UserId = (int)adapter.GetNumber(reader, 7);
            change.UserName = adapter.GetString(reader, 8);
            change.ActionDate = adapter.GetDateTime(reader, 9);

            _sets.Add(change);
        }

        #region IList

        public int Count
        {
            get { return _sets.Count; }
        }

        public bool IsReadOnly
        {
            get { return false; }
        }

        public ChangeRow this[int index]
        {
            get { return _sets[index]; }
            set { _sets[index] = value; }
        }

        public void Add(ChangeRow item)
        {
            _sets.Add(item);
        }

        public void Clear()
        {
            _sets.Clear();
        }

        public bool Contains(ChangeRow item)
        {
            return _sets.Contains(item);
        }

        public void CopyTo(ChangeRow[] array, int arrayIndex)
        {
            _sets.CopyTo(array, arrayIndex);
        }

        public IEnumerator<ChangeRow> GetEnumerator()
        {
            return _sets.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return _sets.GetEnumerator();
        }

        public int IndexOf(ChangeRow item)
        {
            return _sets.IndexOf(item);
        }

        public void Insert(int index, ChangeRow item)
        {
            _sets.Insert(index, item);
        }

        public bool Remove(ChangeRow item)
        {
            return _sets.Remove(item);
        }

        public void RemoveAt(int index)
        {
            _sets.RemoveAt(index);
        }

        #endregion
    }
}
